### PR TITLE
Add SelectLogsWithSigsExcluding and GetLogsWithSigsExcluding

### PR DESCRIPTION
**Pull request by george-dorin**

### CODE REVIEW

File: core/chains/evm/logpoller/log_poller.go [hunk 1]
```diff
@@ -1052,6 +1052,14 @@ func (lp *logPoller) fillRemainingBlocksFromRPC(
...
 }
 
+// GetLogsWithSigsExcluding returns the set difference(A-B) of logs with signature sigA and sigB, matching is done on the topics index
+//
+// For example, query for all unfulfilled requests: request log events which don't have a matching fulfillment log event.
+func (lp *logPoller) GetLogsWithSigsExcluding(ctx context.Context, address common.Address, sigA, sigB common.Hash, topicIndex int, startBlock, endBlock int64, confs int, qopts ...pg.QOpt) (*[]Log, error) {
+	qopts = append(qopts, pg.WithParentCtx(ctx))
+	return lp.orm.SelectLogsWithSigsExcluding(sigA, sigB, topicIndex, address, startBlock, endBlock, confs, qopts...)
+}
+
```
[the remark this answers is not in the document]
Comment (reductionista) [reply core/chains/evm/logpoller/log_poller.go:1070]: (or maybe it should have an Indexed prefix, since this will only work for an indexed topic, as I understand it?)

File: core/chains/evm/logpoller/orm.go [hunk 2]
```diff
@@ -368,3 +368,41 @@ func (o *ORM) SelectIndexedLogs(address common.Address, eventSig common.Hash, to
...
+		AND        a.event_sig = $3
+
+		ORDER BY block_number,log_index ASC
+			`, utils.NewBig(o.chainID), address, sigA.Bytes(), sigB.Bytes(), topicIndex+1, startBlock, endBlock, confs)
+	if err != nil {
+		return nil, err
+	}
+	return &logs, nil
+
+}
```
Comment (reductionista) [inline core/chains/evm/logpoller/orm.go:452]: Query looks good!  Just one thing:  why do we return `&logs` instead of logs?

I notice the other methods all just return type `[]Log`, but this returns `*[]Log`.  Same for the method that calls this... if both of them return `[]Log` instead of `*[]Log`, then the caller won't have to access it by doing stuff like len(*log) or *log[0]... which I notice you do in the tests.

File: core/chains/evm/logpoller/orm_test.go [hunk 3]
```diff
@@ -655,3 +655,133 @@ func BenchmarkLogs(b *testing.B) {
...
+			TxHash:         common.HexToHash("0x0005"),
+			Data:           []byte("requestID-0003 data"),
+		},
+	}))
+	require.NoError(t, orm.InsertBlock(common.HexToHash("0x4"), 4, time.Now()))
+
+	logs, err = orm.SelectLogsWithSigsExcluding(eventSigA, eventSigB, 0, sourceAddr, 0, 4, 0)
+	require.NoError(t, err)
+	require.Len(t, *logs, 1)
+	require.Equal(t, (*logs)[0].Data, []byte("requestID-0002 data"))
```
Comment (reductionista) [inline core/chains/evm/logpoller/orm_test.go:788]: whenever I see *log my instinct it to say we should check to make sure log isn't a nil pointer... in this case, it can't be but we could make that more explicit by returning a slice instead of pointer to slice (see above)

Comment (george-dorin) [reply core/chains/evm/logpoller/orm_test.go]: Fixed in [`aabaab6` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/aabaab6db23df531bc85527260f0c59e54ad56eb)

File: core/chains/evm/logpoller/orm.go (hunk 2 above)
Comment (george-dorin) [reply core/chains/evm/logpoller/orm.go:452]: Fixed in [`aabaab6` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/aabaab6db23df531bc85527260f0c59e54ad56eb)

File: core/chains/evm/logpoller/log_poller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (george-dorin) [reply core/chains/evm/logpoller/log_poller.go:1071]: Indexed seems to make more sense, changed in  [`aabaab6` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/aabaab6db23df531bc85527260f0c59e54ad56eb)

**Review by reductionista — APPROVED** [review]

File: core/chains/evm/logpoller/orm.go [hunk 4]
```diff
@@ -412,3 +412,41 @@ func validateTopicIndex(index int) error {
...
+		AND        a.topics[$5] = b.topics[$5]
+		AND        a.event_sig = $3
+		AND        b.event_sig = $4
+		AND (b.block_number + $8) <= (SELECT COALESCE(block_number, 0) FROM evm_log_poller_blocks WHERE evm_chain_id = $1 ORDER BY block_number DESC LIMIT 1)
+		WHERE      a.evm_chain_id = $1
+		AND        a.address = $2
+		AND        a.event_sig = $3
+
+		ORDER BY block_number,log_index ASC
+			`, utils.NewBig(o.chainID), address, sigA.Bytes(), sigB.Bytes(), topicIndex+1, startBlock, endBlock, confs)
```
[the remark this answers is not in the document]
Comment (george-dorin) [reply core/chains/evm/logpoller/orm.go:450]: Postgres has 1-based arrays so we have to offset it by one to match golang array/slices. That being said it seems I forgot to validate the topicIndex, fixed in [`9360f4b` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/9360f4b78f62efefa495faa5c78e81c793fca64d)

File: core/chains/evm/logpoller/log_poller.go [hunk 5]
```diff
@@ -1060,6 +1061,14 @@ func (lp *logPoller) fillRemainingBlocksFromRPC(
 	return blocksFoundFromRPC, nil
 }
 
+// IndexedLogsWithSigsExcluding returns the set difference(A-B) of logs with signature sigA and sigB, matching is done on the topics index
+//
+// For example, query for all unfulfilled requests: request log events which don't have a matching fulfillment log event.
+func (lp *logPoller) IndexedLogsWithSigsExcluding(ctx context.Context, address common.Address, eventSigA, eventSigB common.Hash, topicIndex int, fromBlock, toBlock int64, confs int, qopts ...pg.QOpt) ([]Log, error) {
```
Comment (connorwstein) [inline core/chains/evm/logpoller/log_poller.go:1067]: hm why the separate context vs just letting callers include their context in qopts like the other methods?

File: core/chains/evm/logpoller/orm_test.go [hunk 6]
```diff
@@ -674,3 +674,133 @@ func BenchmarkLogs(b *testing.B) {
...
+		},
+	}))
+	require.NoError(t, orm.InsertBlock(common.HexToHash("0x4"), 4, time.Now()))
+
+	logs, err = orm.SelectIndexedLogsWithSigsExcluding(eventSigA, eventSigB, 1, sourceAddr, 0, 4, 0)
+	require.NoError(t, err)
+	require.Len(t, logs, 1)
+	require.Equal(t, logs[0].Data, []byte("requestID-0002 data"))
+
+	//No enough confirmations
```
Comment (connorwstein) [inline core/chains/evm/logpoller/orm_test.go:790]: typo

File: core/chains/evm/logpoller/orm_test.go [hunk 7]
```diff
@@ -674,3 +674,133 @@ func BenchmarkLogs(b *testing.B) {
 		require.NoError(b, err)
 	}
 }
+
+func TestSelectLogsWithSigsExcluding(t *testing.T) {
```
Comment (connorwstein) [inline core/chains/evm/logpoller/orm_test.go:678]: lets add some comments throughout this test indicating which cases we're covering here

File: core/chains/evm/logpoller/orm.go [hunk 8]
```diff
@@ -412,3 +412,45 @@ func validateTopicIndex(index int) error {
...
+		SELECT     a.*
+		FROM       evm_logs AS a
+		INNER JOIN evm_logs B
+		ON         a.evm_chain_id = b.evm_chain_id
+		AND        a.address = b.address
+		AND        a.topics[$5] = b.topics[$5]
+		AND        a.event_sig = $3
+		AND        b.event_sig = $4
+		AND (b.block_number + $8) <= (SELECT COALESCE(block_number, 0) FROM evm_log_poller_blocks WHERE evm_chain_id = $1 ORDER BY block_number DESC LIMIT 1)
+		WHERE      a.evm_chain_id = $1
```
Comment (connorwstein) [inline core/chains/evm/logpoller/orm.go:445]: is this where clause necessary ?

File: core/chains/evm/logpoller/orm.go [hunk 9]
```diff
@@ -412,3 +412,45 @@ func validateTopicIndex(index int) error {
...
+		
+		SELECT     a.*
+		FROM       evm_logs AS a
+		INNER JOIN evm_logs B
+		ON         a.evm_chain_id = b.evm_chain_id
+		AND        a.address = b.address
+		AND        a.topics[$5] = b.topics[$5]
+		AND        a.event_sig = $3
+		AND        b.event_sig = $4
+		AND (b.block_number + $8) <= (SELECT COALESCE(block_number, 0) FROM evm_log_poller_blocks WHERE evm_chain_id = $1 ORDER BY block_number DESC LIMIT 1)
```
Comment (connorwstein) [inline core/chains/evm/logpoller/orm.go:444]: hmm so the block range only applies to the first event but the confs apply to both? Wouldn't we want the range to apply to both events as well?

File: core/chains/evm/logpoller/orm_test.go (hunk 7 above)
Comment (connorwstein) [reply core/chains/evm/logpoller/orm_test.go:678]: I don't think this test covers the block range feature or different addresses right?

File: core/chains/evm/logpoller/orm_test.go [hunk 10]
```diff
@@ -674,3 +674,133 @@ func BenchmarkLogs(b *testing.B) {
...
+	require.NoError(t, err)
+	require.Len(t, logs, 1)
+	require.Equal(t, logs[0].Data, []byte("requestID-0002 data"))
+
+	//No enough confirmations
+	logs, err = orm.SelectIndexedLogsWithSigsExcluding(eventSigA, eventSigB, 1, sourceAddr, 0, 5, 5)
+	require.NoError(t, err)
+	require.Len(t, logs, 0)
+
+	require.NoError(t, orm.InsertBlock(common.HexToHash("0x5"), 5, time.Now()))
```
Comment (connorwstein) [inline core/chains/evm/logpoller/orm_test.go:795]: probably want to test the edge cases where say sigA is confirmed by sigB is not and vice versa. Same deal for the block range

**Review by connorwstein — CHANGES_REQUESTED** [review]

File: core/chains/evm/logpoller/orm.go (hunk 9 above)
Comment (reductionista) [reply core/chains/evm/logpoller/orm.go:444]: Seems like we should at least add`block_number >= $6`, which might speed up the query if there's a long history stored.

It's not obvious to me why we'd want to add the upper limit though.  If there's an upper limit for the block range, that seems like it would apply to the request you're looking for.  If it's already been fulfilled after the top of the range you probably don't want it... because you're looking for only unfulfilled requests, right?

Comment (reductionista) [reply core/chains/evm/logpoller/orm.go:444]: I guess there could be more general use cases... probably best to add some clear documentation on how this works, whatever we decide.

File: core/chains/evm/logpoller/log_poller.go (hunk 5 above)
Comment (george-dorin) [reply core/chains/evm/logpoller/log_poller.go]: Removed context in [`e2f04b0` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/e2f04b04bf86327a1c1598d65d2cf4d33cfd0629)

File: core/chains/evm/logpoller/orm_test.go (hunk 6 above)
Comment (george-dorin) [reply core/chains/evm/logpoller/orm_test.go]: Fixed in [`e2f04b0` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/e2f04b04bf86327a1c1598d65d2cf4d33cfd0629)

File: core/chains/evm/logpoller/orm.go (hunk 8 above)
Comment (george-dorin) [reply core/chains/evm/logpoller/orm.go]: Removed in [`e2f04b0` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/e2f04b04bf86327a1c1598d65d2cf4d33cfd0629)

File: core/chains/evm/logpoller/orm.go (hunk 9 above)
Comment (george-dorin) [reply core/chains/evm/logpoller/orm.go:445]: Change in [`e2f04b0` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/e2f04b04bf86327a1c1598d65d2cf4d33cfd0629) so that the block range now applies to both. 
I don't have enough context on the use cases to have a strong opinion on the upper limit but I slightly favor removing it...

File: core/chains/evm/logpoller/orm_test.go (hunk 10 above)
Comment (george-dorin) [reply core/chains/evm/logpoller/orm_test.go:863]: Added in [`e2f04b0` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/e2f04b04bf86327a1c1598d65d2cf4d33cfd0629)

File: core/chains/evm/logpoller/orm_test.go (hunk 7 above)
Comment (george-dorin) [reply core/chains/evm/logpoller/orm_test.go:678]: Added comment and more tests in [`e2f04b0` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/e2f04b04bf86327a1c1598d65d2cf4d33cfd0629)

File: core/chains/evm/logpoller/orm.go (hunk 9 above)
Comment (reductionista) [reply core/chains/evm/logpoller/orm.go:445]: I don't have a strong opinion on whether the upper limit should be there, my point was that if it is, it should apply only to event A, not to both A and B.  So I see the code as having been functionally correct before this change, but incorrect after it.  If we decide it should apply to both of them, we need to document that as it may not be obvious to the user.

Whether the lower limit applies to both events or only to A doesn't affect the result, but having it apply to both can improve performance, so I agree with that part of the change.

Comment (reductionista) [reply core/chains/evm/logpoller/orm.go:445]: Oh, I guess it could affect the result in general, unless we add a clause that ensures the timestamp for event A is earlier than the timestamp for event B.  I'm not sure what the use case would be outside of the request / fulfillment example... maybe I'm thinking too narrowly about the application.

Comment (connorwstein) [reply core/chains/evm/logpoller/orm.go:445]: yeah I also am unsure what future use cases would be, but it does seem possible that a client may want to examine a specific historical range

**Review by connorwstein — APPROVED** [review]

**Review by reductionista — APPROVED** [review]
